crates/subspace-fraud-proof/src/lib.rs

```diff
@@ -29,106 +29,118 @@ use sp_trie::DBValue;
 use std::marker::PhantomData;
 use std::sync::Arc;
 
-/// Returns a storage proof which can be used to reconstruct a partial state trie to re-run
-/// the execution by someone who does not own the whole state.
-pub fn prove_execution<
+/// Creates storage proof for verifying an execution without owning the whole state.
+pub struct ExecutionProver<Block, B, Exec> {
+    backend: Arc<B>,
+    executor: Arc<Exec>,
+    spawn_handle: Box<dyn SpawnNamed>,
+    _phantom: PhantomData<Block>,
+}
+
+impl<Block, B, Exec> ExecutionProver<Block, B, Exec>
+where
     Block: BlockT,
     B: backend::Backend<Block>,
     Exec: CodeExecutor + 'static,
-    Spawn: SpawnNamed + Send + 'static,
-    DB: HashDB<HashFor<Block>, DBValue>,
->(
-    backend: &Arc<B>,
-    executor: &Exec,
-    spawn_handle: Spawn,
-    at: &BlockId<Block>,
-    execution_phase: &ExecutionPhase,
-    delta_changes: Option<(DB, Block::Hash)>,
-) -> sp_blockchain::Result<StorageProof> {
-    let state = backend.state_at(*at)?;
+{
+    /// Constructs a new instance of [`ExecutionProver`].
+    pub fn new(backend: Arc<B>, executor: Arc<Exec>, spawn_handle: Box<dyn SpawnNamed>) -> Self {
+        Self {
+            backend,
+            executor,
+            spawn_handle,
+            _phantom: PhantomData::<Block>,
+        }
+    }
 
-    let trie_backend = state.as_trie_backend().ok_or_else(|| {
-        Box::new(sp_state_machine::ExecutionError::UnableToGenerateProof)
-            as Box<dyn sp_state_machine::Error>
-    })?;
+    /// Returns a storage proof which can be used to reconstruct a partial state trie to re-run
+    /// the execution by someone who does not own the whole state.
+    pub fn prove_execution<DB: HashDB<HashFor<Block>, DBValue>>(
+        &self,
+        at: BlockId<Block>,
+        execution_phase: &ExecutionPhase,
+        delta_changes: Option<(DB, Block::Hash)>,
+    ) -> sp_blockchain::Result<StorageProof> {
+        // TODO: fetch the runtime code properly.
```

Comment:
What do you mean by "properly"? Injected code fetcher?

Reply:
Comment updated https://github.com/subspace/subspace/pull/308/commits/256b421e469f68984c926dbfb14c76a29b60e4d5#diff-f1a9e52b8ec9ebc50e5885824f1d58f7f4ea8bd7158f56ca984dec399a543894R64

Reply:
Hm, why is is necessary to retrieve from primary runtime though? I mean transactions will be executed in secondary runtime, you can as well use it fpr prooving then, can't you?

Reply:
The comment might not be in the right place :D The runtime for proving the execution is the one used for executing the regular tranactions.

Reply:
Then why does it need to change, it seems to be correct already, isn't it?

Reply:
Nope, now the executor node independently loads the runtime code from the local storage, but the secondary runtime code is actually determined by the primary chain, so it should load the execution runtime code from the primary runtime, right?

Reply:
Yes and no. It is determined by the primary node, yes. But it doesn't mean secondary node needs to fetch and compile its runtime on every call. It needs to follow runtime upgrades of the primary node and update its own runtime in its own storage. The rest will work exactly the way it did before. That is how I imagined it at least.

Reply:
That's the implementation detail, and the comment does not imply it will be fetched whenever the runtime code is required :D It just conceptually describes where it should come from originally.

Reply:
It will be addressed in my next PR, so should be fine even it's not crystal clear, I believe you could get the point anyway.